src/rendezvous/channels/MSC3903ECDHv1RendezvousChannel.ts

```diff
@@ -0,0 +1,255 @@
+                    RendezvousFailureReason.UnsupportedAlgorithm,
+                );
+            }
+
+            this.theirPublicKey = decodeBase64(key);
+        } else {
+            // send our public key unencrypted
+            await this.send({
+                algorithm: RendezvousChannelAlgorithm.ECDH_V1,
+                key: encodeBase64(this.ourPublicKey),
+            });
+        }
+
+        this.connected = true;
+
+        this.olmSAS.set_their_key(encodeBase64(this.theirPublicKey!));
+
+        const initiatorKey = isInitiator ? this.ourPublicKey : this.theirPublicKey!;
+        const recipientKey = isInitiator ? this.theirPublicKey! : this.ourPublicKey;
+        let aesInfo = RendezvousChannelAlgorithm.ECDH_V1.toString();
+        aesInfo += `|${encodeBase64(initiatorKey)}`;
+        aesInfo += `|${encodeBase64(recipientKey)}`;
+
+        const aesKeyBytes = this.olmSAS.generate_bytes(aesInfo, 32);
+
+        this.aesKey = await importKey(aesKeyBytes);
+
+        // blank the bytes out to make sure not kept in memory
+        aesKeyBytes.fill(0);
+
+        const rawChecksum = this.olmSAS.generate_bytes(aesInfo, 5);
+        return generateDecimalSas(Array.from(rawChecksum)).join('-');
+    }
+
+    private async encrypt(data: string): Promise<ECDHPayload> {
+        if (!subtleCrypto) {
+            throw new Error('Web Crypto is not available');
+        }
+
+        const iv = new Uint8Array(32);
+        crypto.getRandomValues(iv);
+
+        const encodedData = new TextEncoder().encode(data);
+
+        const ciphertext = await subtleCrypto.encrypt(
+            {
+                name: "AES-GCM",
+                iv,
+                tagLength: 128,
+            },
+            this.aesKey as CryptoKey,
+            encodedData,
+        );
+
+        return {
+            iv: encodeBase64(iv),
+            ciphertext: encodeBase64(ciphertext),
+        };
+    }
+
+    public async send(payload: object) {
+        if (!this.olmSAS) {
+            throw new Error('Channel closed');
+        }
+
+        const data = this.aesKey ? await this.encrypt(JSON.stringify(payload)) : payload;
+
+        await this.transport.send(data);
+    }
+
+    private async decrypt({ iv, ciphertext }: ECDHPayload): Promise<ECDHPayload> {
+        if (!ciphertext || !iv) {
+            throw new Error('Missing ciphertext and/or iv');
+        }
+
+        const ciphertextBytes = decodeBase64(ciphertext);
+
+        if (!subtleCrypto) {
+            throw new Error('Web Crypto is not available');
+        }
+
+        const plaintext = await subtleCrypto.decrypt(
+            {
+                name: "AES-GCM",
+                iv: decodeBase64(iv),
+                tagLength: 128,
+            },
+            this.aesKey as CryptoKey,
+            ciphertextBytes,
+        );
+
+        return JSON.parse(new TextDecoder().decode(new Uint8Array(plaintext)));
+    }
+
+    public async receive(): Promise<object | undefined> {
```

Comment:
This needs better typing, e.g. Generics, see `MatrixEvent::getContent` as an example